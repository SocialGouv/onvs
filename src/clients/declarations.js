import { API_URL } from "@/utils/config"
import fetcher from "@/utils/fetcher"
import { flatObject } from "@/utils/object"

const DECLARATION_ENDPOINT = "declarations"

// Note: this function returns a Promise, as expected, even await is not required by return syntax.
export const createDeclaration = async ({ declaration, keys }) => {
  const data = flatObject(declaration?.steps, keys)
  // Reminder : the id is generated client side, to prevent multiple submits
  data.id = declaration.id
  data.declarationType = declaration.declarationType

  return fetcher(`${API_URL}/${DECLARATION_ENDPOINT}`, {
    body: JSON.stringify(data),
    headers: { "Content-Type": "application/json" },
    method: "POST",
  })
}
export const findDeclaration = async (id) => {
  return fetcher(`${API_URL}/${DECLARATION_ENDPOINT}/${id}`, {
    headers: { "Content-Type": "application/json" },
    method: "GET",
  })
}
