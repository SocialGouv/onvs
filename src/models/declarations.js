import * as yup from "yup"

import * as common from "./common"

// Mapping keys from JS model to DB model
const mappingJStoDB = {
  authors: "authors",
  date: "date",
  declarantContactAgreement: "declarant_contact_agreement",
  declarantEmail: "declarant_email",
  declarantExternalId: "declarant_external_id",
  declarantNames: "declarant_names",
  declarantTel: "declarant_tel",
  declarationType: "declaration_type",
  description: "description",
  factTypes: "fact_types",
  fgDeteriorations: "fg_deteriorations",
  fgGroups: "fg_groups",
  fgOthers: "fg_others",
  // fgOthersPrecision: "fg_others_precision",
  fgStealWithBreakins: "fg_steal_with_breakins",
  fgStealWithoutBreakins: "fg_steal_without_breakins",
  fpDiscriminations: "fp_discriminations",
  fpGroups: "fp_groups",
  fpNoRespects: "fp_no_respects",
  fpOthers: "fp_others",
  // fpOthersPrecision: "fp_others_precision",
  fpPhysicalViolences: "fp_physical_violences",
  // fpHarassments: "fp_harassments",
  fpPhysicalViolencesPrecision: "fp_physical_violences_precision",
  fpPsychologicalViolences: "fp_psychological_violences",
  fpSexualViolences: "fp_sexual_violences",
  fpSpokenViolences: "fp_spoken_violences",
  hour: "hour",
  id: "id",
  job: "job",
  location: "location",
  otherLocation: "other_location",
  pursuit: "pursuit",
  pursuitBy: "pursuit_by",
  rCausePatients: "r_cause_patients",
  rCauseProfessionals: "r_cause_professionals",
  rDeficientCommunications: "r_deficient_communications",
  rDiscords: "r_discords",
  rFalsifications: "r_falsifications",
  rLifeRules: "r_life_rules",
  rNotApparent: "r_not_apparent",
  rOthers: "r_others",
  thirdParty: "third_party",
  town: "town",
  victims: "victims",
}

// Cast the Select input shape ({label, value}) into a simple string
const getValueFromSelect = () =>
  yup.string().transform(function (_, originalValue) {
    return originalValue?.value || ""
  })

const schemaJS = yup.object({
  authors: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  date: yup.string().required(),
  declarantContactAgreement: yup.string(),
  declarantEmail: yup.string(),
  declarantExternalId: yup.string(),
  declarantNames: yup.string(),
  declarantTel: yup.string(),
  declarationType: yup.string(),
  description: yup.string(),
  // factTypes: yup.array().of(yup.string()),
  factTypes: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fgDeteriorations: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fgGroups: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fgOthers: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fgStealWithBreakins: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  fgStealWithoutBreakins: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  fpDiscriminations: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  fpGroups: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  // fpHarassments: yup
  //   .string()
  //   .transform((_, originalValue) => JSON.stringify(originalValue)),
  fpNoRespects: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  fpOthers: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  // fpOthersPrecision: yup.string(),
  fpPhysicalViolences: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  // fgOthersPrecision: yup.string(),
  fpPhysicalViolencesPrecision: yup.string(),
  fpPsychologicalViolences: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fpSexualViolences: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  fpSpokenViolences: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  hour: getValueFromSelect(),
  id: yup.string(),
  job: getValueFromSelect(),
  location: yup.string().required(),
  otherLocation: yup.string(),
  pursuit: yup.string(),
  pursuitBy: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  rCausePatients: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  rCauseProfessionals: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  rDeficientCommunications: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  rDiscords: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  rFalsifications: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),

  rLifeRules: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  rNotApparent: yup.boolean().transform((_, originalValue) => !!originalValue),
  rOthers: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
  thirdParty: yup.string(),
  town: yup.string().required(),
  victims: yup
    .string()
    .transform((_, originalValue) => JSON.stringify(originalValue)),
})

// TODO do the schema in reverse. Beware of rNotApparent, if true, set it to "Pas de motif apparent"
const schemaDB = yup.object({
  r_not_apparent: yup
    .string()
    .transform((_, originalValue) =>
      originalValue ? "Pas de motif apparent" : "",
    ),
})

export const { castJSToDB, castDBToJS, validateJS } = common.build({
  entityName: "declarations",
  mappingJStoDB,
  schemaDB,
  schemaJS,
})

/*
1. validate pour caster et valider
2. renommage des propriétés
*/
