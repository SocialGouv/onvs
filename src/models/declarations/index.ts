import { z } from "zod"
import { Declaration, Prisma } from "@prisma/client"
import prisma from "@/prisma/db"

import {
  ages,
  authorTypes,
  discernmentTroubles,
  etsMainLocations,
  etsSecondaryLocations,
  genders,
  healthJobs,
  isHealthType,
  liberalLocations,
  pursuitComplaintsByValues,
  thirdParties,
  victimTypes,
} from "@/utils/options"
import { difference, uniq } from "lodash"

export enum DeclarationType {
  Liberal = "liberal",
  Ets = "ets",
  Pharmacist = "pharmacist",
}

export enum OuiNonType {
  Oui = "oui",
  Non = "non",
}

export const victimSchema = z
  .object({
    type: z.string().refine((type) => victimTypes.includes(type)),
    gender: z.string().refine((gender) => genders.includes(gender)),
    age: z.string().refine((age) => ages.includes(age)),
    healthJob: z.string().optional(), // the refine rule below will check if this field can be really undefined.
    sickLeaveDays: z.number(),
    hospitalizationDays: z.number(),
    ITTDays: z.number(),
  })
  .strict()
  .refine(
    (val) =>
      isHealthType(val.type)
        ? val.healthJob && healthJobs.includes(val.healthJob)
        : val.healthJob === undefined,
    {
      message:
        "Le champ healthJob doit être rempli (nomenclature healthJobs) si le type est en lien avec une profession de santé (nomenclature healthTypes) et vide sinon.",
    },
  )

export type VictimSchema = z.infer<typeof victimSchema>

export const authorSchema = z
  .object({
    type: z.string().refine((type) => authorTypes.includes(type)),
    gender: z.string().refine((gender) => genders.includes(gender)),
    age: z.string().refine((age) => ages.includes(age)),
    healthJob: z.string().optional(), // the refine rule below will check if this field can be really undefined.
    discernmentTroubles: z
      .string()
      .array()
      .transform((val) => uniq(val))
      .refine((val) => difference(val, discernmentTroubles).length === 0) // Are all the elements included in discernmentTroubles ?
      .optional(),
  })
  .strict()
  .refine(
    (val) =>
      isHealthType(val.type)
        ? val.healthJob && healthJobs.includes(val.healthJob)
        : val.healthJob === undefined,
    {
      message:
        "Le champ healthJob doit être rempli (nomenclature healthJobs) si le type est en lien avec une profession de santé (nomenclature healthTypes) et vide sinon.",
    },
  )

export type AuthorSchema = z.infer<typeof authorSchema>

export const thirdPartySchema = z
  .string()
  .or(z.tuple([z.literal("Autre"), z.string()]))
  .array()
  .transform((val) => uniq(val)) // Remove duplicate entries if any.
  .refine(
    (val) => {
      if (!val.length) return true // Optional is ok.

      // Only 1 array with Autre is allowed.
      const nbArrays = val.filter((elt) => Array.isArray(elt))
      if (nbArrays.length > 1) return false

      for (const elt of val) {
        if (!Array.isArray(elt)) {
          if (!thirdParties.includes(elt)) {
            return false
          }
        }
      }
      return true
    },
    {
      message:
        "Valeur inconnue pour le champ thirdParty ou présence de plusieurs tableaux Autre",
      path: ["thirdParty"],
    },
  )

export type ThirdPartySchemaType = z.infer<typeof thirdPartySchema>

const pursuitSchemaForOther = z
  .object({
    type: z.tuple([z.literal("Autre"), z.string()]),
  })
  .strict()

const pursuitSchemaForLogBook = z
  .object({
    type: z.literal("Main courante"),
  })
  .strict()

const pursuitSchemaForComplaint = z
  .object({
    type: z.literal("Plainte"),
    pursuitBy: z
      .string()
      .array()
      .transform((val) => uniq(val))
      .refine((val) => {
        for (const elt of val) {
          if (!pursuitComplaintsByValues.includes(elt)) return false
        }
        return true
      }),
  })
  .strict()

export const pursuitSchema = z.union([
  pursuitSchemaForOther,
  pursuitSchemaForLogBook,
  pursuitSchemaForComplaint,
])

export type PursuitSchema = z.infer<typeof pursuitSchema>

// Fields available for all declaration types.
const baseSchemaDeclarationApi = z.object({
  id: z.string().uuid().optional(),
  date: z.string(),
  hour: z.string(), // TODO use an enum for hours see Step1.js
  town: z.string(),
  postalCode: z.string().min(4).max(5),
  factTypes: z.string().array(), // TODO check the possible values
  //fpGroups: z.string(), // Not needed, I think.
  fpSpokenViolences: z.string().array(),
  fpPhysicalViolences: z.string().array(),
  fpPhysicalViolencesPrecision: z.string().optional(),
  fpSexualViolences: z.string().array(),
  fpPsychologicalViolences: z.string().array(),
  fpDiscriminations: z.string().array(),
  fpNoRespects: z.string().array(),
  fpOthers: z.string().array(),
  fgGroups: z.string().array(),
  fgDeteriorations: z.string().array(),
  fgStealWithoutBreakins: z.string().array(),
  fgStealWithBreakins: z.string().array(),
  fgOthers: z.string().array(),
  rCausePatients: z.string().array(),
  rCauseProfessionals: z.string().array(),
  rDiscords: z.string().array(),
  rLifeRules: z.string().array(),
  rFalsifications: z.string().array(),
  rDeficientCommunications: z.string().array(),
  rOthers: z.string().array(),
  rOthersPrecision: z.string().optional(),
  rNotApparent: z.boolean(),
  victims: victimSchema.array(),
  authors: authorSchema.array(),
  victims_deprecated: z.object({}).array(),
  authors_deprecated: z.object({}).array(),

  thirdParty: thirdPartySchema.optional(),
  thirdParty_deprecated: z.string().array(),
  thirdPartyIsPresent_deprecated: z.string(), // TODO For now a string, later it should be a boolean.
  thirdPartyPrecision_deprecated: z.string().optional(), // TODO : nest in future thirdParty JSON field.
  pursuit_deprecated: z.string(), // TODO : check over possible values.
  pursuitPrecision_deprecated: z.string().optional(), // TODO : nest in future thirdParty JSON field.
  pursuitBy_deprecated: z.string().array(),

  pursuit: pursuitSchema.optional(),
  description: z.string(),
})

// Specificity for liberal.
const liberalAddonSchema = z.object({
  declarationType: z.literal(DeclarationType.Liberal),
  job: z.string(),
  declarantContactAgreement: z.boolean(),
  declarantContactAgreement_deprecated: z.string(), // TODO: remove when time has come

  declarantNames: z.string().optional(),
  declarantExternalId: z.string().optional(), // TODO : remove this field.
  declarantEmail: z.string().email().optional(),
  declarantTel: z.string().min(10).optional(),

  location_deprecated: z.string(), // TODO remove when time has come
  otherLocation_deprecated: z.string().optional(), // TODO remove when time has come

  location: z
    .object({
      "Dans quel lieu précisément ?": z
        .string()
        .or(z.tuple([z.literal("Autre"), z.string()]))
        .refine(
          (val) => {
            if (!Array.isArray(val)) {
              return liberalLocations.includes(val)
            }
            return true
          },
          {
            message:
              "Les valeurs attendues sont dans la nomenclature liberalLocations",
          },
        ),
    })
    .strict(),
})

// Specificity for ETS.
const etsAddonSchema = z.object({
  declarationType: z.literal(DeclarationType.Ets),
  finesset: z.string().refine(async (finesset) => {
    try {
      const ets = await prisma.ets.findUnique({
        where: {
          finesset,
        },
      })
      console.log("pour ets, on trouve", ets)
      return ets !== null
    } catch (error) {
      console.error("Erreur lors du test du finesset")
      return false
    }
  }),

  locationMain_deprecated: z.string(), // TODO: remove when time has come
  locationSecondary_deprecated: z.string(), // TODO: remove when time has come

  location: z
    .object({
      "Dans quel service ?": z
        .string()
        .refine((val) => etsMainLocations.includes(val), {
          message:
            "Les valeurs attendues sont dans la nomenclature etsMainLocations",
        }),
      "Dans quel lieu précisément ?": z
        .string()
        .refine((val) => etsSecondaryLocations.includes(val), {
          message:
            "Les valeurs attendues sont dans la nomenclature etsSecondaryLocations",
        }),
    })
    .strict(),
})

// TODO : il faut aussi préciser la spécialité en plus du job.
// Specificity for pharmacist.
const pharmacistAddonSchema = z.object({
  declarationType: z.literal(DeclarationType.Liberal),
  job: z.literal("Pharmacien"),

  location: z
    .object({
      "Dans quel lieu précisément ?": z
        .string()
        .or(z.tuple([z.literal("Autre"), z.string()])),
      "Où votre officine se trouve-t-elle ?": z.string(),
      "La pharmacie était-elle ouverte ?": z.nativeEnum(OuiNonType),
      "La pharmacie était-elle de garde ?": z.nativeEnum(OuiNonType),
    })
    .strict(),
})

export const schemaLiberal = baseSchemaDeclarationApi.merge(liberalAddonSchema)
export const schemaEts = baseSchemaDeclarationApi.merge(etsAddonSchema)

export const schemaPharmacist = liberalAddonSchema.merge(pharmacistAddonSchema)

export const schema = z.union([schemaLiberal, schemaEts])

// export type DeclarationApiType = z.infer<typeof DeclarationApiSchema>

export type DeclarationModel = Declaration

export type DeclarationCreateInput = Prisma.DeclarationCreateInput

// export type DeclarationUpdateInput = Prisma.DeclarationUpdateInput

// export type PartialDeclarationModel = {
//   [Property in keyof DeclarationModel]+?: DeclarationModel[Property]
// }
