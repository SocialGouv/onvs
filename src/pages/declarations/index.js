import React from "react"

import { WizardForm } from "@/components/WizardForm"

const DeclarationPage = () => {
  return <WizardForm />
}

export default DeclarationPage
