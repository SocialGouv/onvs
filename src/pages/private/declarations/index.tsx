import Alert from "@/components/Alert"
import Pagination from "@/components/Pagination"
import PrivateLayout from "@/components/PrivateLayout"
import Table from "@/components/Table"
import { FORMAT_DATE } from "@/utils/constants"
import { upperCaseFirstLetters } from "@/utils/string"
import { Declaration } from "@prisma/client"
import { format } from "date-fns"
import Link from "next/link"
import React from "react"
import { BadgeType } from "@/components/BadgeType"
import { useList } from "@/hooks/useList"

function DeclarationAdministration() {
  const [pageIndex, setPageIndex] = React.useState(0)
  const [search, setSearch] = React.useState("")

  const paginatedData = useList<Declaration>({
    url: "/api/declarations",
    pageIndex,
    search,
    setPageIndex,
  })

  const { isLoading, message, list } = paginatedData

  return (
    <PrivateLayout title="Déclarations" leftComponent={null}>
      <Alert message={message} />

      {isLoading ? (
        "Chargement..."
      ) : (
        <Table
          headers={["Date", "Type", "Métier", "Ville", "Description", ""].map(
            (header) => (
              <th
                key={header}
                scope="col"
                className="px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-800 uppercase"
              >
                {header}
              </th>
            ),
          )}
          rows={list?.map((declaration) => (
            <tr key={declaration.id}>
              <td className="px-6 py-4 whitespace-nowrap">
                <div className="text-sm text-gray-900">
                  {declaration?.createdAt
                    ? format(new Date(declaration.createdAt), FORMAT_DATE)
                    : "N/A"}
                </div>
              </td>
              <td className="px-6 py-4 whitespace-nowrap">
                <div className="text-sm text-gray-800">
                  {declaration.declarationType && (
                    <BadgeType type={declaration.declarationType} />
                  )}
                </div>
              </td>
              <td className="px-6 py-4 text-sm text-gray-800 whitespace-nowrap">
                {declaration.job}
              </td>
              <td className="px-6 py-4 text-sm text-gray-800 whitespace-nowrap">
                {upperCaseFirstLetters(declaration.town)}
              </td>
              <td className="px-6 py-4 text-sm text-gray-800">
                <span className="line-clamp-4">{declaration.description}</span>
              </td>
              <td className="px-6 py-4 text-sm font-medium text-right whitespace-nowrap">
                <Link href={`/declaration/${declaration.id}`}>
                  <a className="text-blue-600 hover:text-blue-900">Voir</a>
                </Link>
              </td>
            </tr>
          ))}
        />
      )}
      <Pagination pageIndex={pageIndex} {...paginatedData} />
    </PrivateLayout>
  )
}

export default DeclarationAdministration as React.ReactNode
