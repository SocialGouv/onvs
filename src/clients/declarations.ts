import { DeclarationModel } from "@/models/declarations"
import { API_URL } from "@/utils/config"
import fetcher from "@/utils/fetcher"
import { flatObject } from "@/utils/object"

const DECLARATION_ENDPOINT = "declarations"

// Extract town and postal code in the field. For example : "Nantes (44000)".
function extractTownAndPostalCode(town) {
  // Regex magic ✨
  // There is 2 captured groups : 1/ everthing before a parenthesis, 2/ everything between the '(' and the ')'.
  const groups = town?.match(/([^(]+)\(([^)]+)\)/)

  if (groups?.[2]) {
    return { town: groups[1]?.trim(), postalCode: groups[2]?.trim() }
  }
  return { town }
}

// Note: this function returns a Promise, as expected, even await is not required by return syntax.
export const createDeclaration = async ({
  declaration,
  keys,
}: {
  declaration: any
  keys: Array<any>
}): Promise<string> => {
  const data = flatObject(declaration?.steps, keys)
  const { town, postalCode } = extractTownAndPostalCode(data?.town?.label)

  // console.log("declaration", declaration)
  // console.log("data", data)

  // Reminder : the id is generated client side, to prevent multiple submits.
  data.id = declaration.id
  data.declarationType = declaration.declarationType
  data.town = town
  data.postalCode = postalCode

  data.hour = data.hour.label

  if (declaration?.steps?.job) {
    // job is outside the ordered steps by flow. Refactor the wizard forms to better manage job info transfer
    data.job = declaration.steps.job.job.label
  }

  // LOCATION ---------------------------------------------------------------

  if (data.declarationType === "ets") {
    // Filling the legacy columns (renamed in _deprecated to distinguish them).
    if (data.locationMain?.label) {
      data.locationMain_deprecated = data.locationMain.label
    }

    if (data.locationSecondary?.label) {
      data.locationSecondary_deprecated = data.locationSecondary.label
    }

    // Filling the new column.
    data.location = {
      "Dans quel service ?": data.locationMain?.label,
      "Dans quel lieu précisément ?": data.locationSecondary?.label,
    }

    // We need to remove these fields since they are renamed in _deprecated.
    delete data.locationMain
    delete data.locationSecondary
  } else {
    // Should be the liberal flow.
    data.location_deprecated = data.location
    data.otherLocation_deprecated = data.otherLocation

    // Fill the new column.
    data.location = {
      "Dans quel lieu précisément ?": data.otherLocation
        ? ["Autre", data.otherLocation]
        : data.location,
    }

    // We need to remove the newly deprecated fied as zod expect not to see it.
    delete data.otherLocation
  }

  // DECLARANT_CONTACT_AGREEMENT ---------------------------------------------------------------
  // Filling the legacy column for user agreement (renamed in _deprecated).
  data.declarantContactAgreement_deprecated = data.declarantContactAgreement
  // Filling the new column for user agreement (which is of boolean type and can be null)
  data.declarantContactAgreement =
    data.declarantContactAgreement_deprecated === "true"
      ? true
      : data.declarantContactAgreement_deprecated === "false"
      ? false
      : null

  // TODO: the victims and authors shouldn't be with the { label, value } shape in db.
  // Do a migration to reshape the data in db (prod) and make the following to adapt the client.

  // let victims = data.victims

  // if (victims?.length) {
  //   victims = victims.map((victim) => {
  //     return {
  //       ...victim,
  //       type: victim.type.label,
  //       gender: victim.gender.label,
  //       age: victim.age.label,
  //     }
  //   })

  //   data.victims = victims
  // }

  // let authors = data.authors

  // if (authors?.length) {
  //   authors = authors.map((author) => {
  //     return {
  //       ...author,
  //       type: author.type.label,
  //       gender: author.gender.label,
  //       age: author.age.label,
  //     }
  //   })

  //   data.authors = authors
  // }

  // console.log("dans clients", data)

  return fetcher(`${API_URL}/${DECLARATION_ENDPOINT}`, {
    body: JSON.stringify(data),
    headers: { "Content-Type": "application/json" },
    method: "POST",
  })
}
export const findDeclaration = async (
  id: string,
): Promise<DeclarationModel> => {
  return fetcher(`${API_URL}/${DECLARATION_ENDPOINT}/${id}`, {
    headers: { "Content-Type": "application/json" },
    method: "GET",
  })
}
