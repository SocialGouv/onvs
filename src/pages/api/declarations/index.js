import Cors from "micro-cors"

import { create } from "@/services/declarations"

import prisma from "@/prisma/db"
import { buildMetaPagination } from "@/utils/pagination"
import { handleErrors, handleNotAllowedMethods } from "@/utils/api"
import { DuplicateError } from "@/utils/errors"

const UNIQUE_VIOLATION_PG = "23505"

const handler = async (req, res) => {
  res.setHeader("Content-Type", "application/json")

  try {
    switch (req.method) {
      case "GET": {
        const totalCount = await prisma.declaration.count()

        const { pageIndex, pageSize, totalPages, prismaQueryParams } =
          await buildMetaPagination({ totalCount, ...req.query })

        const declarations = await prisma.declaration.findMany({
          orderBy: [{ createdAt: "desc" }],
          ...prismaQueryParams,
        })

        return res
          .status(200)
          .json({ declarations, pageIndex, totalCount, pageSize, totalPages })
      }

      case "POST": {
        const id = await create(req.body)

        return res.status(200).json({ id })
      }
      default: {
        handleNotAllowedMethods(req, res)
      }
    }
  } catch (error) {
    if (error?.code === UNIQUE_VIOLATION_PG) {
      // eslint-disable-next-line no-ex-assign
      error = new DuplicateError(error.message)
      error.statusCode = 409
    }

    handleErrors(error, res)
  }
}

const cors = Cors({
  allowMethods: ["GET", "OPTIONS", "POST"],
})

export default cors(handler)
