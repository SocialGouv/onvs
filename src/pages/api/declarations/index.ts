import Cors from "micro-cors"
import { NextApiRequest, NextApiResponse } from "next"
import { z } from "zod"
import { pipe } from "lodash/fp"

import prisma from "@/prisma/db"
import { create } from "@/services/declarations"
import withSession from "@/lib/session"
import { UserLoggedModel } from "@/models/users"
import { buildMetaPagination } from "@/utils/pagination"
import { checkAllowedMethods, handleApiError } from "@/utils/api"
import {
  AuthenticationError,
  AuthorizationError,
  DuplicateError,
} from "@/utils/errors"
import { jobsByOrders } from "@/utils/options"
import { getEditorFromToken } from "@/services/editors"

function buildWhereClause(user: UserLoggedModel) {
  switch (user.role) {
    case "Gestionnaire d'ordre": {
      if (!user?.scope?.order) throw new Error("This is not supposed to happen")
      return {
        job: {
          in: jobsByOrders[user.scope.order],
        },
      }
    }
    case "Gestionnaire Ã©tablissement": {
      if (!user?.scope?.ets) throw new Error("This is not supposed to happen")
      return {
        finesset: {
          in: user.scope.ets,
        },
      }
    }
  }
}

const handler = async (
  req: NextApiRequest & { session: any },
  res: NextApiResponse,
) => {
  res.setHeader("Content-Type", "application/json")

  switch (req.method) {
    case "GET": {
      const user = req.session.get("user")

      if (!user?.isLoggedIn) {
        throw new AuthenticationError()
      }
      const whereClause = buildWhereClause(user)

      const totalCount = await prisma.declaration.count({
        where: whereClause,
      })

      const { pageIndex, pageSize, totalPages, prismaPaginationQueryParams } =
        await buildMetaPagination({
          totalCount,
          ...req.query,
        })

      const declarations = await prisma.declaration.findMany({
        orderBy: [{ createdAt: "desc" }],
        ...prismaPaginationQueryParams,
        where: whereClause,
      })

      return res.status(200).json({
        data: declarations,
        pageIndex,
        totalCount,
        pageSize,
        totalPages,
      })
    }

    case "POST": {
      const bearerError = new AuthorizationError(
        "A header 'Authorization: Bearer <editor-id>' must be present.",
      )

      const invalidTokenError = new AuthorizationError(
        "The provided token is not valid.",
      )

      if (
        !req.headers.authorization ||
        !/Bearer/.test(req.headers.authorization)
      )
        throw bearerError

      const token = req.headers.authorization.replace("Bearer", "").trim()

      if (!token) throw bearerError

      const schemaGuid = z.string().uuid()

      if (!schemaGuid.safeParse(token)?.success) {
        throw invalidTokenError
      }

      const editor = await getEditorFromToken(token)

      if (!editor) throw invalidTokenError

      try {
        const id = await create(req.body, editor)
        return res.status(200).json({ id })
      } catch (error) {
        console.error("Error Prisma ", error.message)
        if (error?.code === "P2002")
          throw new DuplicateError("A declaration with this id already exists.")
        else throw error
      }
    }
  }
}

const allowMethods = ["GET", "POST"]

export default pipe(
  withSession,
  Cors({
    allowMethods,
  }),
  checkAllowedMethods({ allowMethods }),
  handleApiError,
)(handler)
