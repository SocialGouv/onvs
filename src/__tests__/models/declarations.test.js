import { castJSToDB, validateJS } from "@/models/declarations/liberal";

import declarationSample from "./declaration-sample.json";

test("A correct declaration", async () => {
  const declaration = {
    date: "2020-09-09",
    job: {
      label: "Chiropracteur",
      value: "Chiropracteur",
    },
    location: "Cabinet individuel",
    otherLocation: "",
    town: "vincennes",
  };

  await expect(validateJS(declaration)).resolves.toMatchInlineSnapshot(`
          Object {
            "date": "2020-09-09",
            "job": "Chiropracteur",
            "location": "Cabinet individuel",
            "otherLocation": "",
            "town": "vincennes",
          }
        `);
});

test("A incorrect declaration", async () => {
  const consoleErrorLegacy = console.error;
  console.error = jest.fn();

  const declaration = {
    date: "",
    job: "Étudiant en santé",
    location: "Cabinet individuel",
    otherLocation: "",
    town: "vincennes",
  };

  await expect(validateJS(declaration)).rejects.toMatchInlineSnapshot(
    `[Error: Données invalides  (declarations/liberal modèle)]`
  );
  console.error = consoleErrorLegacy;
});

test("A full correct declaration", async () => {
  await expect(castJSToDB(declarationSample?.form)).resolves.toMatchSnapshot();
});
