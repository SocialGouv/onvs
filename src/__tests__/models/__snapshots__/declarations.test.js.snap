// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A full correct declaration 1`] = `
Object {
  "authors": "[{\\"discernmentTroubles\\":[\\"Prise d’alcool\\"],\\"type\\":{\\"label\\":\\"Inconnu\\",\\"value\\":\\"Inconnu\\"},\\"gender\\":{\\"label\\":\\"Féminin\\",\\"value\\":\\"Féminin\\"},\\"age\\":{\\"label\\":\\"+ de 18 ans\\",\\"value\\":\\"+ de 18 ans\\"}}]",
  "date": "2020-09-09",
  "declarant_contact_agreement": "true",
  "declarant_email": "marie.dupuis@gmail.com",
  "declarant_external_id": "12345678901",
  "declarant_names": "mom",
  "declarant_tel": "0202020202",
  "description": "Voici ce qui s'est passé...",
  "fact_types": "[\\"Atteinte aux personnes\\",\\"Atteinte aux biens\\"]",
  "fg_deteriorations": "[]",
  "fg_groups": "[\\"Vol avec effraction\\",\\"Autres faits\\"]",
  "fg_others": "[\\"Escroquerie\\"]",
  "fg_steal_with_breakins": "[\\"Informations\\",\\"Vol à main armée\\"]",
  "fg_steal_without_breakins": "[]",
  "fp_discriminations": "[]",
  "fp_groups": "[\\"Les auteurs n’ont pas respecté les règles du lieu / ont eu un comportement incivique\\",\\"Autres faits\\"]",
  "fp_no_respects": "[\\"Nuisance, chahut, fugue\\",\\"Consommation ou détention sur place d’alcool et/ou de produits stupéfiants pour son propre usage sexuel\\"]",
  "fp_others": "[\\"Atteinte au respect dû aux morts\\"]",
  "fp_physical_violences": "[]",
  "fp_psychological_violences": "[]",
  "fp_sexual_violences": "[]",
  "fp_spoken_violences": "[]",
  "hour": "Matin (7h-12h)",
  "job": "Chiropracteur",
  "location": "Cabinet individuel",
  "pursuit": "Plainte",
  "pursuit_by": "[\\"L'ordre\\"]",
  "r_cause_patients": "[\\"De paiement\\"]",
  "r_cause_professionals": "[\\"De soins\\"]",
  "r_deficient_communications": "[]",
  "r_discords": "[\\"Entre les professionnels\\"]",
  "r_falsifications": "[]",
  "r_life_rules": "[]",
  "r_others": "[\\"Autre\\"]",
  "third_party": "Autre",
  "town": "vincennes",
  "victims": "[{\\"type\\":{\\"label\\":\\"Étudiant en santé\\",\\"value\\":\\"Étudiant en santé\\"},\\"gender\\":{\\"label\\":\\"Masculin\\",\\"value\\":\\"Masculin\\"},\\"age\\":{\\"label\\":\\"- de 18 ans\\",\\"value\\":\\"- de 18 ans\\"},\\"sickLeaveDays\\":1,\\"hospitalizationDays\\":0,\\"ITTDays\\":0,\\"healthJob\\":{\\"label\\":\\"Ambulancier\\",\\"value\\":\\"Ambulancier\\"}},{\\"type\\":{\\"label\\":\\"Détenu\\",\\"value\\":\\"Détenu\\"},\\"gender\\":{\\"label\\":\\"Féminin\\",\\"value\\":\\"Féminin\\"},\\"age\\":{\\"label\\":\\"- de 18 ans\\",\\"value\\":\\"- de 18 ans\\"},\\"sickLeaveDays\\":1,\\"hospitalizationDays\\":0,\\"ITTDays\\":0}]",
}
`;
